File: 0_100/046_Permutations/Solution.py
class Solution:
    
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        arr = []
        self.backtrace(nums, arr)
        return self.ans
    
    def backtrace(self, nums, arr):
        if len(nums) == len(arr):
            self.ans.append([i for i in arr])
        else:
            for i in nums:
                if i not in arr:
                    arr.append(i)
                    self.backtrace(nums, arr)
                    arr.pop()

File: 0_100/046_Permutations/Solution.rs
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans: Vec<Vec<i32>> = vec![];
        let mut arr: Vec<i32> = vec![];
        Solution::backtrace(&nums, &mut arr, &mut ans);
        return ans;
    }

    fn backtrace(nums: &Vec<i32>, arr: &mut Vec<i32>, ans: &mut Vec<Vec<i32>>) {
        if arr.len() == nums.len() {
            ans.push(arr.clone());
        } else {
            for i in nums {
                if !arr.contains(i) {
                    arr.push(*i);
                    Solution::backtrace(nums, arr, ans);
                    arr.pop();
                }
            }
        }
    }
}